Project/DAQ_System/Program/Src/Sensor/Accelerometer/LSM6DSOXAccelerometer.cpp

```diff
@@ -0,0 +1,176 @@
+    double acc_magnitude = 0;
+    double average_array[kSampleCount];
+
+    // Collects samples to calibrate for 1G
+    for(int i = 0; i < kSampleCount; i++) {
+        ReadRawAcceleration();
+        acc_magnitude = sqrt(raw_acceleration_data_[0] * raw_acceleration_data_[0] + raw_acceleration_data_[1] * raw_acceleration_data_[1] + raw_acceleration_data_[2] * raw_acceleration_data_[2]);
+        average_array[i] = acc_magnitude * 0.001;
+    }
+
+    double min_value = std::numeric_limits<double>::max();
+    double max_value = -1;
+
+    // Find the largest and smallest magnitudes
+    for(int j = 2; j < kSampleCount; j++) {
+        if(average_array[j] <= min_value) {
+            min_value = average_array[j];
+
+        } else if (average_array[j] > max_value) {
+            max_value = average_array[j];
+        }
+    }
+
+    // Check that the maximum deviation is below a certain threshold.
+    // If it is too large, the conversion factor will be set to a previously
+    // determined conversion factor
+    gravity_adjustment_conversion_factor_ = 0;
+
+    if (max_value - min_value < 0.09) {
+        double total_array = 0;
+	    for(int i = 0; i < kSampleCount; i++) {
+		    total_array += average_array[i];
+	    }
+
+	    gravity_adjustment_conversion_factor_ = total_array / kSampleCount;
+    } else {
+        // Testing found that ~1.028 is about 1G
+        gravity_adjustment_conversion_factor_ = 1.028;
+    }
+
+}
+
+void LSM6DSOX::SetODR(SensorConfiguration::ODR ODRValue){
+	static constexpr uint8_t kNumBytes = 2;
+    uint8_t commands[kNumBytes] = {0};
+    commands[0] = CTRL1_XL;
+    uint8_t currentRegisterValue[1] = {0};
+
+    // read the current register values for the CTRL_X register
+	HAL_I2C_Master_Transmit(&i2c_,ACC_ADDRESS, commands,1, HAL_MAX_DELAY);
+	HAL_I2C_Master_Receive(&i2c_,ACC_ADDRESS, currentRegisterValue,1, HAL_MAX_DELAY);
+
+	// set the ODR bits to the desired value without touching the other bits in the register
+	// ODRValue is shifter 4 bits to properly align the desired ODR bits to the proper bits in the register
+	commands[1] = (currentRegisterValue[0] & 0x0F) | ODRValue << 4;
+
+	// write to register with the desired bit values
+	HAL_I2C_Master_Transmit(&i2c_,ACC_ADDRESS, commands,kNumBytes, HAL_MAX_DELAY);
+
+}
+
+void LSM6DSOX::SetFSR(SensorConfiguration::FSR FSRValue){
+	static constexpr uint8_t kNumBytes = 2;
+    uint8_t commands[kNumBytes] = {0};
+    commands[0] = CTRL1_XL;
+    uint8_t currentRegisterValue[1] = {0};
+
+    // read the current register values for the CTRL1_XL register
+	HAL_I2C_Master_Transmit(&i2c_,ACC_ADDRESS, commands,1, HAL_MAX_DELAY);
+	HAL_I2C_Master_Receive(&i2c_,ACC_ADDRESS, currentRegisterValue,1, HAL_MAX_DELAY);
+
+	// set the FSR bits to the desired value without touching the other bits in the register
+	// FSRValue is shifted to properly align the FSR bits in to the register.
+	commands[1] = (currentRegisterValue[0] & 0xF3) | FSRValue << 2  ;
```

Comment:
One problem that occurred when trying to make both an ODR and FSR function is that they appear on the same register. This meant that I had to be careful when writing to the same register multiple times because I didn't want to overwrite previous changes to the register. So in order to do that I had to use logical operators on the current value of the register to mask bits and set bits so that I could maintain the previous register values and just change the bits that I wanted to. 

![odr6 66k](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/113472670/f26d2c5d-9ee9-4950-8772-cfbb76526b13) 

in this first picture you can see that I am changing the ODR to 6.66Khz which is 1010 in the 4 MSB of the register. bit 5 and 6 are the FSR bits and needed to be written without accidentally changing the values of the first 4 MSB
![16gFSR](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/113472670/ead1648e-d42b-44f1-9322-49af5808bb74)

This next picture shows that happening as I write an FSR value of 16g which is 01 to bits 5 and 6 while still maintaining the 6.66Khz ODR that I previously wrote.

Reply:
> One problem that occurred when trying to make both an ODR and FSR function is that they appear on the same register.

As long as you are not clearing the value in the register, it should be retained. Perhaps clear it at initialization. Otherwise, for configuring the sensor, using only the logical OR `|` should be sufficient.